axum-extra/src/middleware/opentelemetry.rs

```diff
@@ -0,0 +1,480 @@
+/// pipeline sending traces to jaeger.
+///
+/// [otel]: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/http.md
+/// [`Router::into_make_service_with_connect_info`]: axum::Router::into_make_service_with_connect_info
+pub fn opentelemetry_tracing_layer() -> TraceLayer<
+    SharedClassifier<ServerErrorsAsFailures>,
+    OtelMakeSpan,
+    OtelOnRequest,
+    OtelOnResponse,
+    OtelOnBodyChunk,
+    OtelOnEos,
+    OtelOnFailure,
+> {
+    TraceLayer::new_for_http()
+        .make_span_with(OtelMakeSpan)
+        .on_request(OtelOnRequest)
+        .on_response(OtelOnResponse)
+        .on_body_chunk(OtelOnBodyChunk)
+        .on_eos(OtelOnEos)
+        .on_failure(OtelOnFailure)
+}
+
+/// A [`MakeSpan`] that creates tracing spans using [OpenTelemetry's conventional field names][otel].
+///
+/// [otel]: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/http.md
+#[derive(Clone, Copy, Debug)]
+pub struct OtelMakeSpan;
+
+impl<B> MakeSpan<B> for OtelMakeSpan {
+    fn make_span(&mut self, req: &Request<B>) -> Span {
+        let user_agent = req
+            .headers()
+            .get(header::USER_AGENT)
+            .map_or("", |h| h.to_str().unwrap_or(""));
+
+        let host = req
+            .headers()
+            .get(header::HOST)
+            .map_or("", |h| h.to_str().unwrap_or(""));
+
+        let scheme = req
+            .uri()
+            .scheme()
+            .map_or_else(|| "HTTP".into(), http_scheme);
+
+        let http_route = if let Some(matched_path) = req.extensions().get::<MatchedPath>() {
+            matched_path.as_str().to_owned()
+        } else if let Some(uri) = req.extensions().get::<OriginalUri>() {
+            uri.0.path().to_owned()
+        } else {
+            req.uri().path().to_owned()
+        };
+
+        let uri = if let Some(uri) = req.extensions().get::<OriginalUri>() {
+            uri.0.clone()
+        } else {
+            req.uri().clone()
+        };
+        let http_target = uri
+            .path_and_query()
+            .map(|path_and_query| path_and_query.to_string())
+            .unwrap_or_else(|| uri.path().to_owned());
+
+        let client_ip = parse_x_forwarded_for(req.headers())
+            .or_else(|| {
+                req.extensions()
+                    .get::<ConnectInfo<SocketAddr>>()
+                    .map(|ConnectInfo(client_ip)| Cow::from(client_ip.to_string()))
+            })
+            .unwrap_or_default();
+
+        let remote_context = extract_remote_context(req.headers());
+        let remote_span = remote_context.span();
+        let span_context = remote_span.span_context();
+        let trace_id = span_context
+            .is_valid()
+            .then(|| Cow::from(span_context.trace_id().to_string()))
+            .unwrap_or_default();
+
+        let span = tracing::info_span!(
+            "HTTP request",
```

Comment:
This is going to form the "name" of the span. We do want it to be relatively low cardinality (as per the [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/http.md#name), but the generally accepted form on the server side is something like `{http.method} {http.route}`, presuming `route` had a matched path (so _only_ if `MatchedPath` came back as `Some`). In the event we don't have a matched path, then this should just be something like the `HTTP request` that you have here or `HTTP {http.method} request`.

That keeps the cardinality in check but also makes the span more easilly identifiable.

Reply:
AFAIK non-constant values cannot be used to name a tracing span. I've got around this on my tracing layer by using opentelemetry spans instead.

Reply:
You can set the name with the format `{http.method} {http.route}` to `otel.name`, I tested it with the otel exporter going to jaeger afterwards and I get the otel.name for the root span instead of `HTTP Request`